projects/validation-messages/src/lib/components/validation-messages/validation-messages.component.ts

```diff
@@ -142,74 +116,105 @@ export class ValidationMessagesComponent
     );
   }
 
-  ngOnInit(): void {
-    this.readFormControlByControlName();
-  }
-
-  ngAfterContentInit(): void {
-    this.readFormControlFromHost();
-
-    this.control.valueChanges
-      .pipe(takeUntil(this.ngUnsubscribe))
-      .subscribe(() => this.cd.markForCheck());
-  }
-
-  ngOnDestroy(): void {
-    this.ngUnsubscribe.next();
-    this.ngUnsubscribe.complete();
-  }
-
-  ngDoCheck(): void {
-    if (
-      this.control &&
-      ((this.control.invalid && this.control.touched) ||
-        (!this.control.invalid && this.shownErrors.length > 0))
-    ) {
-      this.updateErrorMessages();
-      this.cd.markForCheck();
-    }
-  }
-
-  private updateErrorMessages(): void {
-    this.shownErrors = [];
+  private _updateErrorMessages(): void {
+    this.shownErrors.set([]);
 
     if (!this.control || !this.control.errors) {
       return;
     }
 
     const controlErrors = this.control.errors;
+    let errorCounter = 0;
     for (const propertyName in controlErrors) {
-      if (!this.multiple && this.shownErrors.length === 1) {
+      if (!this._multiple() && errorCounter === 1) {
         break;
       }
 
       if (controlErrors[propertyName] && propertyName !== 'server') {
-        this.shownErrors.push(
+        const errorMessage =
           this.validationMessagesService.getValidatorErrorMessage(
             propertyName,
             controlErrors[propertyName],
             this.errorsMessages
-          )
-        );
+          );
+
+        this.shownErrors.update((errors) => [...errors, errorMessage]);
+        errorCounter++;
       }
     }
   }
+  private _valueChangesEffect() {
+    effect(
+      () => {
+        this.controlValueChanges();
+
+        if (
+          this.control &&
+          ((this.control.invalid && this.control.dirty) ||
+            (!this.control.invalid && this.shownErrors.length > 0))
+        ) {
+          this._updateErrorMessages();
+        }
+      },
+      {
+        allowSignalWrites: true,
+      }
+    );
+  }
+
+  private _apiErrorMessagesEffect() {
+    effect(() => {
+      this._clearErrors();
+      const errorMessages = this._apiErrorMessages();
+      this.parseApiErrorMessages(errorMessages);
+      this.showServerErrors = true;
+
+      if (this.control && errorMessages) {
+        this.control.setErrors({
+          server: errorMessages,
+        });
+      }
+    });
+  }
 
-  private unsubscribeAndClearValueChanges(): void {
+  private _multipleEffect() {
+    effect(
+      () => {
+        this._multiple();
+        this._updateErrorMessages();
+      },
+      {
+        allowSignalWrites: true,
+      }
+    );
+  }
```

Comment:
can `this._updateErrorMessages();` be performed in the `multiple` setter, so we could get rid of this effect?